### PR TITLE
Add new Combinatorics node (WIP)

### PR DESCRIPTION
**Pull request by DolphinDream**
The combinatoric operations implemented so far (based on itertools module) are:

PRODUCT
PERMUTATIONS
COMBINATIONS

The node is vectorized and works with lists of any type.

- [x] Code changes complete.
- [x] Code documentation complete.
- [x] Documentation for users complete (or not required, if user never sees these changes).
- [x] Manual testing done. 
- [ ] Unit-tests implemented.
- [x] Ready for merge.

Node UI:
sv-combinatoricsnode-ui1

Product operation:
sv-combinatorics-product

note: The example shows the product of 3 different lists (numbers, strings and vertices).

Permutations operation (with vectorization):
sv-combinatorics-permutation

Combinations operation (with vectorization):
sv-combinatorics-combination

### CODE REVIEW

File: nodes/list_mutators/combinatorics.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    def sv_init(self, context):
+        self.inputs.new('StringsSocket', "Repeat").prop_name = "repeat"
+        self.inputs.new('StringsSocket', "Length").prop_name = "length"
+        self.inputs.new('StringsSocket', "Selector")
+        self.inputs.new('StringsSocket', "A")
+        self.inputs.new('StringsSocket', "B")
+
+        self.outputs.new('StringsSocket', "Result")
+
+        self.operation = "PRODUCT"
```
Comment (zeffii) [inline nodes/list_mutators/combinatorics.py]: this is already set in the EnumProperty default. If you are doing this just to trigger update_operation, just call it straight up.
```diff
- self.operation = "PRODUCT"
+ self.update_operation(context)
```
if you really need to.

Comment (DolphinDream) [reply nodes/list_mutators/combinatorics.py]: good point. I did add the default value to the enum after I had added the above code in sv_init :)

**Comment by DolphinDream** [comment]
And here’s a fun application.. the node speaks fluent binary ;)

sv-combinatorics-product-binarycode

**Comment by DolphinDream** [comment]
And another fun (practical) application.. 3D grid:

well. i’m sure there are other ways to create this ouptut... however, the real fun part begins beyond 3D :) Obviously this setup can also create 4D grids (which is why I dived into developing this node first since it is useful for the developing of the quaternions / 4D math nodes 😉 )

sv-combinatorics-3d-grid

**Comment by DolphinDream** [comment]
If there are no objections, comments, suggestions.. this node should be ready to land.

**Comment by DolphinDream** [comment]
Btw, currently the node is placed in the BETA section of the index.md... (though its node/doc file location is in List Mutator). If appropriate I can move it to the List Mutator in the index as well, unless it is recommended to keep it in beta  for now until it is further beta tested by users.

File: nodes/list_mutators/combinatorics.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        self.inputs.new('StringsSocket', "B")
+
+        self.outputs.new('StringsSocket', "Result")
+
+        self.update_operation(context)
+
+    def update(self):
+        ''' Add/remove sockets as A-Z sockets are connected/disconnected '''
+
+        # not a multiple quaternion operation ? => no need to update sockets
```
Comment (zeffii) [inline nodes/list_mutators/combinatorics.py]: which quaternions?!

Comment (DolphinDream) [reply nodes/list_mutators/combinatorics.py]: oops. cut and paste :)

**Pull request merged**